src/pixelator/analysis/polarization/__init__.py

```diff
@@ -99,60 +103,113 @@ def polarization_scores_component(
         )
         return pd.DataFrame()
 
-    # create antibody node counts
-    counts_df = create_node_markers_counts(graph=graph, k=0)
+    def morans(graph: Graph, n_permutations):
+        N = float(graph.vcount())
 
-    # remove markers with zero variance
-    counts_df = counts_df.loc[
-        :, (counts_df != 0).any(axis=0) & (counts_df.nunique() > 1)
-    ]
+        W = graph.get_adjacency_sparse()
+        # Normalize the weights by the degree of the node
+        W = W / W.sum(axis=0)
+        X = create_node_markers_counts(graph)
 
-    # clr transformation
-    if normalization == "clr":
-        counts_df = clr_transformation(df=counts_df, non_negative=True, axis=0)
+        # remove markers with zero variance
+        X = X.loc[:, (X != 0).any(axis=0) & (X.nunique() > 1)]
 
-    # compute the spatial weights matrix (w) from the graph
-    w = WSP(graph.get_adjacency_sparse()).to_W(silence_warnings=True)
+        # Calculate normalization factor
+        C = N / W.sum()
 
-    # compute polarization statistics for each marker using the spatial weights (w)
-    # and the markers counts distribution (y) (Morans I autocorrelation)
-    def data():
-        for m in counts_df.columns:
-            yield morans_autocorr(w, counts_df[m], permutations)
+        X_perm = [
+            clr_transformation(x) if normalization == "clr" else x
+            for x in permutations(X, n=n_permutations, random_seed=random_seed)
+        ]
+
+        if normalization == "clr":
+            X = clr_transformation(X, non_negative=True, axis=0)
 
-    if permutations > 0:
-        df = pd.DataFrame(
-            ((m.I, m.p_rand, m.z_rand, m.p_sim, m.z_sim) for m in data()),
+        _compute_morans_i_per_marker = partial(
+            _compute_morans_i,
+            marker_count_matrix=X,
+            marker_count_matrix_permuted=X_perm,
+            weight_matrix=W,
+            normalization_factor=C,
+        )
+        markers = X.columns
+
+        results = pd.DataFrame(
+            map(_compute_morans_i_per_marker, markers),
             columns=[
+                "marker",
                 "morans_i",
-                "morans_p_value",
                 "morans_z",
-                "morans_p_value_sim",
-                "morans_z_sim",
-            ],
-        ).fillna(0)
-    else:
-        df = pd.DataFrame(
-            ((m.I, m.p_rand, m.z_rand) for m in data()),
-            columns=[
-                "morans_i",
+                "perm_mean",
+                "perm_std",
                 "morans_p_value",
-                "morans_z",
             ],
-        ).fillna(0)
+        )
+        return results
+
+    start_time = time.perf_counter()
+    logger.info("Computing Moran's I for component: %s", component_id)
+
+    results = morans(graph, n_permutations)
```

Comment:
Yeah. This is a good point, and I see why it's confusing. I will skip passing an argument to this function since it is all relaying on being the the closure anyway. The reason to encapsulate it in a function here is that I think it makes the code slightly easier to read.